### PR TITLE
Make sure haInfo sync map is not invoked on an event-loop

### LINKED ISSUE #1302: EventLoop thread blocked on Vertx.clusteredVertx()
Vertx 3.2.1.
Sometimes EventLoop thread blocked on Vertx.clusteredVertx().
Method Vertx.clusteredVertx() called on the main java thread on application startup.

Stacktrace:

```
19:30:48.633 [vertx-blocked-thread-checker] WARN  i.v.core.impl.BlockedThreadChecker - Thread Thread[vert.x-eventloop-thread-0,5,main] has been blocked for 5660 ms, time limit is 2000
io.vertx.core.VertxException: Thread blocked
    at java.lang.Object.wait(Native Method) ~[na:1.8.0]
    at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.pollResponse(InvocationFuture.java:299) ~[hazelcast-3.5.2.jar:3.5.2]
    at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.waitForResponse(InvocationFuture.java:247) ~[hazelcast-3.5.2.jar:3.5.2]
    at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.get(InvocationFuture.java:224) ~[hazelcast-3.5.2.jar:3.5.2]
    at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.get(InvocationFuture.java:204) ~[hazelcast-3.5.2.jar:3.5.2]
    at com.hazelcast.map.impl.proxy.MapProxySupport.invokeOperation(MapProxySupport.java:456) ~[hazelcast-3.5.2.jar:3.5.2]
    at com.hazelcast.map.impl.proxy.MapProxySupport.putInternal(MapProxySupport.java:417) ~[hazelcast-3.5.2.jar:3.5.2]
    at com.hazelcast.map.impl.proxy.MapProxyImpl.put(MapProxyImpl.java:97) ~[hazelcast-3.5.2.jar:3.5.2]
    at com.hazelcast.map.impl.proxy.MapProxyImpl.put(MapProxyImpl.java:87) ~[hazelcast-3.5.2.jar:3.5.2]
    at io.vertx.core.impl.HAManager.<init>(HAManager.java:145) ~[vertx-core-3.2.1.jar:na]
    at io.vertx.core.impl.VertxImpl.lambda$new$270(VertxImpl.java:152) ~[vertx-core-3.2.1.jar:na]
    at io.vertx.core.impl.VertxImpl$$Lambda$3/852445367.handle(Unknown Source) ~[na:na]
    at io.vertx.core.impl.FutureImpl.checkCallHandler(FutureImpl.java:158) ~[vertx-core-3.2.1.jar:na]
    at io.vertx.core.impl.FutureImpl.setHandler(FutureImpl.java:100) ~[vertx-core-3.2.1.jar:na]
    at io.vertx.core.impl.ContextImpl.lambda$null$16(ContextImpl.java:305) ~[vertx-core-3.2.1.jar:na]
    at io.vertx.core.impl.ContextImpl$$Lambda$7/299713639.handle(Unknown Source) ~[na:na]
    at io.vertx.core.impl.ContextImpl.lambda$wrapTask$18(ContextImpl.java:335) ~[vertx-core-3.2.1.jar:na]
    at io.vertx.core.impl.ContextImpl$$Lambda$8/977225900.run(Unknown Source) ~[na:na]
    at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:358) ~[netty-common-4.0.33.Final.jar:4.0.33.Final]
    at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:357) ~[netty-transport-4.0.33.Final.jar:4.0.33.Final]
    at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:112) ~[netty-common-4.0.33.Final.jar:4.0.33.Final]
    at java.lang.Thread.run(Thread.java:744) ~[na:1.8.0]
```

### PR DESCRIPTION
**Pull request by tsegismont**
Fixes #1302

Most often, code in HAManager is invoked from cluster manager threads.
Some parts though need to be called from a worker thread.

### CODE REVIEW

**Comment by tsegismont** [comment]
HA tests pass on Vert.x core, currently running Hazelcast CM. Will try others as well.

**Comment by tsegismont** [comment]
@vietj clustering test suite runs fine with Hazelcast and Infinispan cluster managers. On other CMs, I can see the usual intermittent failures.

File: src/main/java/io/vertx/core/impl/VertxImpl.java [hunk 1]
```diff
@@ -668,10 +683,18 @@ public void undeploy(String deploymentID) {
 
   @Override
   public void undeploy(String deploymentID, Handler<AsyncResult<Void>> completionHandler) {
-    if (haManager() != null && haManager().isEnabled()) {
-      haManager().removeFromHA(deploymentID);
-    }
-    deploymentManager.undeployVerticle(deploymentID, completionHandler);
+    this.executeBlocking(fut -> {
+      if (haManager() != null && haManager().isEnabled()) {
```
Comment (vietj) [inline src/main/java/io/vertx/core/impl/VertxImpl.java]: move the condition outside of the executeBlocking block so we don't execute the blocking task if there is no need to be. In addition you can capture the haManager in a local variable.

File: src/main/java/io/vertx/core/impl/VertxImpl.java [hunk 2]
```diff
@@ -525,40 +536,44 @@ public synchronized void close(Handler<AsyncResult<Void>> completionHandler) {
...
-
-              int serverCount = httpServers.size() + netServers.size();
-
-              AtomicInteger serverCloseCount = new AtomicInteger();
-
-              Handler<AsyncResult<Void>> serverCloseHandler = res -> {
-                if (res.failed()) {
-                  log.error("Failure in shutting down server", res.cause());
+        this.executeBlocking(fut -> {
+          if (haManager() != null) {
```
Comment (vietj) [inline src/main/java/io/vertx/core/impl/VertxImpl.java]: move the test outside the task so we don't execute the task when there is no need to be

**Comment by tsegismont** [comment]
@vietj done

**Pull request merged**